### PR TITLE
refactor and test: addressing external audit

### PR DESCRIPTION
**Pull request by kupermind**
- Addressing external audit.

### CODE REVIEW

File: contracts/MemeArbitrum.sol [hunk 1]
```diff
@@ -9,14 +9,22 @@ interface IWETH {
 
 // @title MemeArbitrum - a smart contract factory for Meme Token creation on Arbitrum.
 contract MemeArbitrum is MemeFactory {
+    // Base UniswapV3 pool cardinality that corresponds to 240 seconds window (240 / 1 seconds per block)
+    uint16 public constant POOL_CARDINALITY = 240;
```
Comment (kupermind) [inline contracts/MemeArbitrum.sol:13]: This affects gas usage a lot. Have communicated with auditors for these new numbers across all the networks:
240 for Arbitrum
120 for Base
60 for Celo and ETH

File: contracts/MemeBase.sol [hunk 2]
```diff
@@ -57,7 +64,7 @@ contract MemeBase is MemeFactory {
         uint256 totalAmount;
         for (uint256 i = 0; i < accounts.length; ++i) {
             totalAmount += amounts[i];
-            memeHearters[launchCampaignNonce][accounts[i]] = amounts[i];
+            memeHearters[launchCampaignNonce][accounts[i]] += amounts[i];
```
Comment (kupermind) [inline contracts/MemeBase.sol:67]: Issue with launch account amount increments

File: contracts/MemeArbitrum.sol [hunk 3]
```diff
@@ -9,14 +9,22 @@ interface IWETH {
...
+
     /// @dev MemeArbitrum constructor
     constructor(
-        address _olas,
```
Comment (kupermind) [inline contracts/MemeArbitrum.sol:14]: Issue with `olas` address not needed in factories.

File: contracts/MemeFactory.sol [hunk 4]
```diff
@@ -183,10 +177,10 @@ abstract contract MemeFactory {
             : (memeToken, nativeToken, memeTokenAmount, nativeTokenAmount);
 
         // Calculate the price ratio (amount1 / amount0) scaled by 1e18 to avoid floating point issues
-        uint256 priceX96 = (amount1 * 1e18) / amount0;
+        uint256 price = FixedPointMathLib.divWadDown(amount1, amount0);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:180]: Changing to equivalent line and correcting the variable name

File: contracts/MemeFactory.sol [hunk 5]
```diff
@@ -220,13 +214,13 @@ abstract contract MemeFactory {
 
         // Schedule for ascendance leftovers from native token
         // Note that meme token leftovers will be purged via purgeThisMeme
-        uint256 nativeLeftovers = isNativeFirst ? (nativeTokenAmount - amount0) : (nativeTokenAmount - amount1);
+        uint256 nativeLeftovers = nativeTokenAmount - (isNativeFirst ? amount0 : amount1);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:217]: As suggested - a bit neater

File: contracts/MemeFactory.sol [hunk 6]
```diff
@@ -220,13 +214,13 @@ abstract contract MemeFactory {
...
         if (nativeLeftovers > 0) {
             scheduledForAscendance += nativeLeftovers;
         }
 
         // Increase observation cardinality
-        IUniswapV3(pool).increaseObservationCardinalityNext(60);
+        IUniswapV3(pool).increaseObservationCardinalityNext(_observationCardinalityNext());
```
Comment (kupermind) [inline contracts/MemeFactory.sol:223]: As proposed by auditors

File: contracts/MemeFactory.sol [hunk 7]
```diff
@@ -250,16 +244,8 @@ abstract contract MemeFactory {
...
-        uint256 memeAmountToBurn;
-
-        if (isNativeFirst) {
-            nativeAmountForOLASBurn = amount0;
-            memeAmountToBurn = amount1;
-        } else {
-            memeAmountToBurn = amount0;
-            nativeAmountForOLASBurn = amount1;
-        }
+        (uint256 nativeAmountForOLASBurn, uint256 memeAmountToBurn) = isNativeFirst ? (amount0, amount1) :
```
Comment (kupermind) [inline contracts/MemeFactory.sol:247]: As suggested - a bit neater

File: contracts/MemeFactory.sol [hunk 8]
```diff
@@ -312,16 +298,16 @@ abstract contract MemeFactory {
     ) internal returns (address memeToken) {
         bytes32 randomNonce = keccak256(abi.encodePacked(block.timestamp, msg.sender, memeNonce));
         randomNonce = keccak256(abi.encodePacked(randomNonce));
-        bytes memory payload = abi.encodePacked(type(Meme).creationCode, abi.encode(name, symbol, DECIMALS, totalSupply));
-        // solhint-disable-next-line no-inline-assembly
-        assembly {
-            memeToken := create2(0x0, add(0x20, payload), mload(payload), randomNonce)
-        }
+        memeToken = address(new Meme{salt: randomNonce}(name, symbol, DECIMALS, totalSupply));
```
Comment (kupermind) [inline contracts/MemeFactory.sol:301]: As suggested - have a high level call

File: contracts/MemeFactory.sol [hunk 9]
```diff
@@ -436,9 +421,7 @@ abstract contract MemeFactory {
         require(memeSummon.unleashTime == 0, "Meme already unleashed");
 
         // Update meme token map values
-        uint256 totalNativeTokenCommitted = memeSummon.nativeTokenContributed;
-        totalNativeTokenCommitted += msg.value;
-        memeSummon.nativeTokenContributed = totalNativeTokenCommitted;
+        memeSummon.nativeTokenContributed += msg.value;
```
Comment (kupermind) [inline contracts/MemeFactory.sol:424]: As suggested - no need to have more than 1 line

File: contracts/MemeFactory.sol [hunk 10]
```diff
@@ -581,7 +564,7 @@ abstract contract MemeFactory {
 
     /// @dev Collects all accumulated LP fees.
     /// @param tokens List of tokens to be iterated over.
-    function collectFees(address[] memory tokens) public {
+    function collectFees(address[] memory tokens) external {
```
Comment (kupermind) [inline contracts/MemeFactory.sol:567]: As suggested - this needs to be external

File: contracts/MemeFactory.sol [hunk 11]
```diff
@@ -600,4 +583,10 @@ abstract contract MemeFactory {
 
         _locked = 1;
     }
+
+    /// @dev Gets global number of unleashed meme tokens.
+    /// @return Number of unleashed meme tokens.
+    function numTokens() external view returns (uint256) {
```
Comment (kupermind) [inline contracts/MemeFactory.sol:589]: As suggested - having a public getter

**Review by 77ph — APPROVED** [review]

File: contracts/MemeFactory.sol [hunk 12]
```diff
@@ -364,7 +350,6 @@ abstract contract MemeFactory {
 
         // Push token into the global list of tokens
         memeTokens.push(memeToken);
-        numTokens = memeTokens.length;
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:367]: ok

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**